lib/ddtrace/span/external_resource.rb

```diff
@@ -0,0 +1,38 @@
+require 'ddtrace/ext/app_types'
+require 'ddtrace/ext/http'
+
+module Datadog
+  class Span
+    # Extension for Datadog::Span that tracks if this span
+    # represents a unit of work performed by this application
+    # or an external resource being indirectly measured
+    # (e.g. SQL, Redis, outbound HTTP request).
+    module ExternalResource
+      attr_writer :external_resource
+
+      # Span types of known to represent an internal application trace
+      INTERNAL_APPLICATION_SPAN_TYPES = [
+        Datadog::Ext::AppTypes::CUSTOM,
+        Datadog::Ext::AppTypes::WEB,
+        Datadog::Ext::AppTypes::WORKER,
+        Datadog::Ext::HTTP::TEMPLATE
+      ].freeze
+
+      # Computes if this span represents an external resource by
+      # first checking if +span.external_resource+ was explicitly set.
+      # If so, use that value.
+      #
+      # Else, check if +span.span_type+ is a span type known to
+      # represent internal application work.
+      # If so, return +false+, otherwise return +true+.
+      #
+      # @return [Boolean] true if this span represents work performed outside of this application
+      def external_resource?
+        return @external_resource if defined?(@external_resource)
+        return false unless span_type
+
+        !INTERNAL_APPLICATION_SPAN_TYPES.include?(span_type)
```

Comment:
I think I see a bit of an existential dilemma here in determining whether span is external exclusively by this span type: if there is any intersection between the types of operations done in the internal domain and the external domain, then this strategy would likely mislabel the span.

For example, I could see a `custom` span that represents an external service e.g. wrapping an invocation of a Bash script or command line utility from Ruby.

I don't think this strategy is flexible enough and will require `Span` to possess a lot of architectural knowledge about "what is external" and what is not; I don't think this is an appropriate responsibility for `Span`, even when encapsulated in this module.

Instead, I would suggest a more explicit approach: `Span` gains an ` external_resource` boolean field from this module, then the instrumentation manually changes this with `trace.trace { |span| span.external_resource = true }`. By default it would be `false`.

This would 1) simplify the responsibilities of `Span` to just being a struct (which I think is a general design direction we should maintain as much as possible), 2) move the responsibilities to the instrumentation which already has the responsibility to describe the measurement in question, which 3) consequently puts this decision in the hands of the user/auto-instrumentation which is the most authoritative source of truth for this domain knowledge.